modules/end-user-guide/partials/proc_building-custom-intellij-editor-images.adoc

```diff
@@ -0,0 +1,151 @@
+[id="building-custom-intellij-editor-images_{context}"]
```

Comment:
[id="building-custom-intellij-editor-images_{context}"]
= Building custom IntelliJ editor images

This procedure describes how to build an IntelliJ-based editor image using a custom archive and a custom image name.

.Prerequisites

* Docker 18.09 or greater that supports link:https://docs.docker.com/develop/develop-images/build_enhancements/[Docker BuildKit]
* `git`
* `jq`
* Java Development Kit (JDK) link:https://docs.oracle.com/en/java/javase/11/install/installation-jdk-linux-platforms.html[version 11]
* `wget`
* On the macOS platform: `+gnu-getopt+`

.Procedure

. Obtain the download URL for the IntelliJ IDEA archive:

.. Go to link:https://www.jetbrains.com/idea/download/other.html[Download IntelliJ IDEA Other Versions].

.. Scroll to version 2020.3 or older. (Currently, more recent versions than 2020.3 are not supported by Projector.)

.. Locate the URL for downloading the `+tar.gz+` archive of the custom IntelliJ-based distribution for Linux with bundled JBR.

.. Copy the URL for a subsequent step.

. Define the following locations for publishing the editor image and metadata:

.. The location of the IntelliJ editor image in a registry accessible by {prod-short}: `__<IntelliJ-container-image>__`

.. The HTTPS resource containing the editor metadata: `__<URL_to_meta.yaml>__`

. Define the build parameters.

.. Specify the tag to be applied to the image after the build: `--tag __<name:tag>__`

.. Paste the download URL for the IntelliJ IDEA archive from the first step: `--url __<downloadUrl>__`

. Get a local copy of the link:https://github.com/che-incubator/jetbrains-editor-images[JetBrains Projector Editor Images repository].
+
[source]
----
$ git clone https://github.com/che-incubator/jetbrains-editor-images
$ cd jetbrains-editor-images
----

. Build the container.
+
[subs="+quotes,macros,attributes"]
----
$ ./projector build --tag __<name:tag>__ --url __<downloadUrl>__
----
+
.Building the Docker image with WebStorm 2020.3.3
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh build --tag che-webstorm:latest --url https://download.jetbrains.com/webstorm/WebStorm-2020.3.3.tar.gz
----
====
+
.Building the Docker image with PyCharm Community 2020.3.5
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh build --tag che-pycharm:latest --url https://download.jetbrains.com/python/pycharm-community-2020.3.5.tar.gz
----
====
+
.Building the Docker image with IntelliJ IDEA Ultimate 2020.2.2
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh build --tag che-idea-ultimate:latest --url https://download.jetbrains.com/idea/ideaIU-2020.2.2.tar.gz
----
====

. To test the container, run the container locally and go to `++http://localhost:8887++` to access the JetBrains IDE.
+
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run __<name:tag>__
----
+
.Testing the Docker image with WebStorm
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run che-webstorm:latest
----
====
+
.Testing the Docker image with PyCharm
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run che-pycharm:latest
----
====
+
.Testing the Docker image with IntelliJ IDEA Ultimate
====
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run che-idea-ultimate:latest
----
====

. Publish the container to `__<IntelliJ-container-image>__`.

. Create a `+meta.yaml+` file with the following content:
+
[source,yaml,subs="+quotes,macros,attributes"]
----
apiVersion: v2
publisher: <username>
name: intellij-ide
version: latest
type: Che Editor
displayName: IntelliJ-based IDE
title: IntelliJ-based IDE as Editor for Eclipse Che
description: IntelliJ-based IDE running using Projector
icon: https://www.jetbrains.com/apple-touch-icon.png
category: Editor
repository: https://github.com/che-incubator/jetbrains-editor-images
firstPublicationDate: "2021-04-10"
spec:
  endpoints:
   -  name: "intellij"
      public: true
      targetPort: 8887
      attributes:
        protocol: http
        type: ide
        path: /projector-client/index.html?backgroundColor=434343&wss
  containers:
   - name: intellij-ide
     image: "__<IntelliJ-container-image>__"
     mountSources: true
     volumes:
         - mountPath: "/home/projector-user"
           name: projector-user
     ports:
         - exposedPort: 8887
     memoryLimit: "4096M"
----

. Publish the `+meta.yaml+` file to `__<URL_to_meta.yaml>__`.

.Additional resources

* xref:using-intellij-based-ides.adoc[]
* link:https://github.com/che-incubator/jetbrains-editor-images[JetBrains Projector Editor Images repository]
* link:https://docs.docker.com/develop/develop-images/build_enhancements/[Docker BuildKit]